rust.003.vectorized/src/main.rs

```diff
@@ -1,18 +1,118 @@
 #![feature(core_intrinsics)]
+#[cfg(target_arch = "x86")]
+use std::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+
 use std::fs::File;
 use std::io::{BufRead, BufReader};
 
 use std::intrinsics::unlikely;
 
+const CHUNKSIZE: usize = 32;
+const BUFSIZE: usize = 4096;
+
+/// Sum each adjacent 8bit lane into u64 x 4
+#[inline]
+unsafe fn hsum_epu8_epu64(v: __m256i) -> __m256i {
+    _mm256_sad_epu8(v, _mm256_setzero_si256())
+}
+
+/// Sum u64 x 4 -> i64
+#[inline]
+unsafe fn hsum_epu64_scalar(v: __m256i) -> i64 {
+    let lo = _mm256_castsi256_si128(v); // get the lower 8 u16s
+    let hi = _mm256_extracti128_si256(v, 1); // get the upper 8 u16
+    let sum2x64 = _mm_add_epi64(lo, hi); // narrow to 128
+
+    let hi = _mm_unpackhi_epi64(sum2x64, sum2x64);
+    let sum = _mm_add_epi64(hi, sum2x64); // narrow to 64
+    _mm_cvtsi128_si64(sum)
+}
+
+/// Count AT / GT occurances in buffer.
+/// Relies on the fact that `G|C` & 10 == 2, `A|T` & 10 == 0, and `N` & 10 == 10.
+/// Ideally the buffershould be a multiple of CHUNKSIZE
+#[inline]
+fn count_gc_at(buffer: &[u8]) -> (i64, i64) {
```

Comment:
Of note, this is agnostic to the line size or any other restriction on sizes / overflows.